File: backend/app/schemas/notice.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import datetime
from app.models.enums import NoticeTypeEnum, NoticeStatusEnum

# NoticeTemplate Schemas
class NoticeTemplateBase(BaseModel):
    notice_type: NoticeTypeEnum
    version: int
    content_template: str
    is_active: bool = True

class NoticeTemplateCreate(NoticeTemplateBase):
    pass

class NoticeTemplateResponse(NoticeTemplateBase):
    id: str
    signed_off_by_name: Optional[str] = None
    signed_off_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

class NoticeTemplateSignOff(BaseModel):
    signed_off_by_name: str

# Notice Schemas
class NoticeBase(BaseModel):
    case_id: str
    target_account_id: Optional[str] = None
    notice_type: NoticeTypeEnum
    status: NoticeStatusEnum = NoticeStatusEnum.DRAFTED
    recipient_bank_name: Optional[str] = None
    recipient_nodal_email: Optional[str] = None
    recipient_bank_ifsc: Optional[str] = None
    supersedes_notice_id: Optional[str] = None

class NoticeCreate(NoticeBase):
    pass

class NoticeResponse(NoticeBase):
    id: str
    notice_number: str
    template_version: Optional[int] = None
    pdf_file_path: Optional[str] = None
    issued_by_user_id: Optional[str] = None
    sent_at: Optional[datetime] = None
    sla_deadline_at: Optional[datetime] = None
    responded_at: Optional[datetime] = None
    response_summary: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

class NoticeStatusUpdate(BaseModel):
    status: NoticeStatusEnum
    response_summary: Optional[str] = None

File: frontend/src/api/notice.ts
import { api } from "./client";

export type NoticeTypeEnum = "section_94" | "section_168" | "section_106" | "unfreeze_order" | "clarification";
export type NoticeStatusEnum = "drafted" | "sent" | "acknowledged" | "action_taken" | "rejected" | "clarification_requested";

export interface NoticeTemplate {
  id: string;
  notice_type: NoticeTypeEnum;
  version: number;
  content_template: string;
  is_active: boolean;
  signed_off_by_name?: string;
  signed_off_at?: string;
  created_at: string;
  updated_at: string;
}

export interface NoticeCreate {
  case_id: string;
  target_account_id?: string;
  notice_type: NoticeTypeEnum;
  status?: NoticeStatusEnum;
  recipient_bank_name?: string;
  recipient_nodal_email?: string;
  recipient_bank_ifsc?: string;
  supersedes_notice_id?: string;
}

export interface Notice {
  id: string;
  notice_number: string;
  case_id: string;
  target_account_id?: string;
  notice_type: NoticeTypeEnum;
  status: NoticeStatusEnum;
  recipient_bank_name?: string;
  recipient_nodal_email?: string;
  recipient_bank_ifsc?: string;
  template_version?: number;
  pdf_file_path?: string;
  issued_by_user_id?: string;
  sent_at?: string;
  sla_deadline_at?: string;
  responded_at?: string;
  response_summary?: string;
  supersedes_notice_id?: string;
  created_at: string;
  updated_at: string;
}

export const noticeService = {
  getTemplates: async (): Promise<NoticeTemplate[]> => {
    const response = await api.get("/notices/templates");
    return response.data;
  },

  createTemplate: async (data: Partial<NoticeTemplate>): Promise<NoticeTemplate> => {
    const response = await api.post("/notices/templates", data);
    return response.data;
  },

  signOffTemplate: async (id: string, name: string): Promise<NoticeTemplate> => {
    const response = await api.put(`/notices/templates/${id}/sign-off`, { signed_off_by_name: name });
    return response.data;
  },

  generateNotice: async (data: NoticeCreate): Promise<Notice> => {
    const response = await api.post("/notices/generate", data);
    return response.data;
  },

  getNoticesForCase: async (caseId: string): Promise<Notice[]> => {
    const response = await api.get(`/notices/case/${caseId}`);
    return response.data;
  },

  updateNoticeStatus: async (id: string, status: NoticeStatusEnum, summary?: string): Promise<Notice> => {
    const response = await api.put(`/notices/${id}/status`, { status, response_summary: summary });
    return response.data;
  },

  getDownloadUrl: (noticeId: string): string => {
    return `${api.defaults.baseURL}/notices/${noticeId}/download`;
  },

  downloadNotice: async (noticeId: string): Promise<void> => {
    const response = await api.get(`/notices/${noticeId}/download`, {
      responseType: "blob",
    });
    const disposition = response.headers["content-disposition"] as string | undefined;
    let filename = `notice-${noticeId}.pdf`;
    if (disposition) {
      const match = disposition.match(/filename="?([^";\n]+)"?/);
      if (match?.[1]) filename = match[1];
    }
    const blob = new Blob([response.data], { type: "application/pdf" });
    const url = window.URL.createObjectURL(blob);
    const anchor = document.createElement("a");
    anchor.href = url;
    anchor.download = filename;
    document.body.appendChild(anchor);
    anchor.click();
    anchor.remove();
    window.URL.revokeObjectURL(url);
  },

  downloadPack: async (noticeId: string): Promise<void> => {
    const response = await api.get(`/notices/${noticeId}/pack`, {
      responseType: "blob",
    });
    const disposition = response.headers["content-disposition"] as string | undefined;
    let filename = `notice-pack-${noticeId}.zip`;
    if (disposition) {
      const match = disposition.match(/filename="?([^";\n]+)"?/);
      if (match?.[1]) filename = match[1];
    }
    const blob = new Blob([response.data], { type: "application/zip" });
    const url = window.URL.createObjectURL(blob);
    const anchor = document.createElement("a");
    anchor.href = url;
    anchor.download = filename;
    document.body.appendChild(anchor);
    anchor.click();
    anchor.remove();
    window.URL.revokeObjectURL(url);
  },
};
